firedrake/slate/slac/utils.py

```diff
@@ -136,6 +152,224 @@ def visit_Symbol(self, o, *args, **kwargs):
         return SymbolWithFuncallIndexing(o.symbol, o.rank, o.offset)
 
 
+# singledispatch for class functions
+def classsingledispatch(func):
+    dispatcher = singledispatch(func)
+
+    def wrapper(*args, **kw):
+        return dispatcher.dispatch(args[1].__class__)(*args, **kw)
+
+    wrapper.register = dispatcher.register
+    update_wrapper(wrapper, func)
+    return wrapper
+
+
+class SlateTranslator():
+    """Multifunction for translating UFL -> GEM.  """
+
+    def __init__(self, builder):
+        self.builder = builder
+        self.decomp_dict = OrderedDict()
+
+    def slate_to_gem_translate(self):
+        translated_nodes = OrderedDict()
+        traversed_dag = list(post_traversal([self.builder.expression]))
+
+        # First traversal for resolving tensors and assembled vectors
+        for tensor in traversed_dag:  # tensor hier is actually TensorBase
+            if isinstance(tensor, sl.Tensor) or isinstance(tensor, sl.AssembledVector):
+                translated_nodes.setdefault(tensor, self.slate_to_gem(tensor, translated_nodes))
+
+        # Second traversal for other nodes
+        for tensor in traversed_dag[:len(traversed_dag)-1]:
+            # other tensor types are translated into gem nodes
+            if not isinstance(tensor, sl.Tensor) and not isinstance(tensor, sl.AssembledVector):
+                translated_nodes.setdefault(tensor, self.slate_to_gem(tensor, translated_nodes))
```

Comment:
Why do you need to split these loops? as in first `Tensor` and `AssembledVector`, and then the rest.  Post-order traversal should suffice to ensure that the dependencies are processed first...

Reply:
Right. Adressed in 418eaea8e307de39983cf9eb60caa8e01d266212.